README.md
# JWT

## JWT是什么

`JSON Web Token (JWT)`是一个开放标准(RFC 7519)，它定义了一种紧凑的、自包含的方式，用于作为JSON对象在各方之间安全地传输信息。该信息可以被验证和信任，因为它是数字签名的。

## 什么时候你应该用JWT

* Authorization (授权) : 这是使用JWT的最常见场景。一旦用户登录，后续每个请求都将包含JWT，允许用户访问该令牌允许的路由、服务和资源。单点登录是现在广泛使用的JWT的一个特性，因为它的开销很小，并且可以轻松地跨域使用。

* Information Exchange (信息交换) : 对于安全的在各方之间传输信息而言，JSON Web Tokens无疑是一种很好的方式。因为JWT可以被签名，例如，用公钥/私钥对，你可以确定发送人就是它们所说的那个人。另外，由于签名是使用头和有效负载计算的，您还可以验证内容没有被篡改。

## 使用

1. 安装依赖

```ts
yarn add yarn add @yunflyjs/yunfly-plugin-fast-jwt
```

2. config/config.plugin.ts 中声明插件

```ts filename="config/config.plugin.ts" {6-9}
/**
 * yunfly 插件
 * 数组顺序就是插件的加载顺序
 */
const plugins: {[key:string]: string}[] = [
  {
    name: 'jwt',
    package: '@yunflyjs/yunfly-plugin-fast-jwt'
  }
];
// 
export default plugins;
```

3. config/config.default.ts 中启用插件 config.jwt

```js filename="src/config/config.default.ts"
config.jwt = {
  enable: true,
  expiredPassThrough: false,
  secret:'YUNFLYJS_JWT_TOKEN_DEMO',
  expire: '1h',
  token: { 
    type: 'cookie', 
    key: 'Authorization', 
    httpOnly: true, 
    domain:'127.0.0.1', 
    path:'/',
    // 20分钟后过期
    expires: () => { return new Date(Date.now() + 60 * 1000 * 20) }, 
  },
  // token: { type:'header', key: 'Authorization' },
  rsSign: {
    enable: false,
    interval: 15,
  },
  unless: ['/favicon.ico'],
  global: true,
  passThrough: false,
}
```

* `jwt.config` 还可以是一个函数

```ts filename="src/config/config.default.ts"
type JWTConfig = JWTOptions | ((ctx: Context) => JWTOptions)

config.jwt = (ctx: Context) => {
  return {
    enable: true,
    expiredPassThrough: false,
    secret:'YUNFLYJS_JWT_TOKEN_DEMO',
    expire: '1h',
    ......
  }
}
```

## 配置说明

| 参数 | 类型 | 默认值 |必填 | 说明 |
| ------ | ------ | ------ | ------ | ------ |
| enable | `boolean` |  `true`  | 是 | 是否启用 JWT 校验 |
| secret | `string` |  `YUNFLYJS_JWT_TOKEN_DEMO` | 是 | JWT 盐值 |
| expiredPassThrough | `boolean` |  `false`  | 否 | 当`JWT`过期时是`自动重签`还是向外`抛出过期错误`。自动重签会一直不过期，抛出过期错误需要自定义错误处理逻辑 |
| expire | `string \| number`| `3h` |  否  | 过期时间 Eg: 60, "2 days", "10h", "7d".  ("120" is equal to "120ms"). |
| token | `JWTTokenOptions` |  `{ type:'header', key: 'Authorization' }` |  否  | token配置项 (重要参数，请关注下面说明。)|
| rsSign | `RsSignOptions` |  `{ enable: false, interval: 15 }` |  否  |自动续签,`interval`单位为分，默认为15分钟。若开启，当判断过期时间是否快超过`interval`分钟， 如果快超过就重新生成JWT秘钥。若`expire`小于等于`interval`则不处理|
| global | `boolean` | `true` | 否 |   是否全局启用 JWT 校验 (重要参数，请关注下面说明。) |
| unless | `string[]` | `['/favicon.ico']` | 否  | { global:true }时，此参数可用，需要排除 JWT 验证的接口 |
| passThrough | `boolean` | `false`| 否  | { global:true }时，此参数可用，校验不通过时是否继续执行 |

## token 参数说明

token 表示 JWT 参数传递方式，目前有两种方式进行传递

* 1、 通过 `header` 头进行校验
* 2、 通过 `cookie` 进行校验

| 参数 | 类型 | 必填 | 说明 |
| ------ | ------ | ------ | ------ |
| type | `header | cookie` | 是 | 参数传递方式 |
| key | `string` | 是 | JWT 校验的 key 值 |
| httpOnly | `boolean` | 否 | type:cookie 时有效 |
| domain | `boolean` | 否 | type:cookie 时有效 |
| path | `string[]` | 否 | type:cookie 时有效 |
| maxAge | `number` | 否 | type:cookie 时有效, 一个数字, 表示从 Date.now() 得到的毫秒数 |
| expires | `Date \| ()=>Date` | 否 | type:cookie 时有效, 一个 Date 对象, 表示 cookie 的到期日期 (默认情况下在会话结束时过期) 或者[函数，返回一个Data对象] |
| secure | `boolean` | 否 | type:cookie 时有效, 一个布尔值, 表示 cookie 是否仅通过 HTTPS 发送 (HTTP 下默认为 false, HTTPS 下默认为 true) |
| sameSite | `boolean  \|  string` | 否 | type:cookie 时有效, 一个布尔值或字符串, 表示该 cookie 是否为 "相同站点" cookie (默认为 false). 可以设置为 'strict', 'lax', 'none', 或 true (映射为 'strict') |
| signed | `boolean` | 否 | type:cookie 时有效, 一个布尔值, 表示是否要对 cookie 进行签名 (默认为 false). 如果为 true, 则还会发送另一个后缀为 .sig 的同名 cookie, 使用一个 27-byte url-safe base64 SHA1 值来表示针对第一个 Keygrip 键的 cookie-name=cookie-value 的哈希值. 此签名密钥用于检测下次接收 cookie 时的篡改 |
| overwrite | `boolean` | 否 | type:cookie 时有效, 一个布尔值, 表示是否覆盖以前设置的同名的 cookie (默认是 false). 如果是 true, 在同一个请求中设置相同名称的所有 Cookie（无论路径或域）是否在设置此Cookie 时从 Set-Cookie 消息头中过滤掉 |

## global 参数说明

`global` 是一个比较重要的参数，参数类型为 `boolean`

* `global = true`
当`global`值为`true`时，则表示当前`node服务`所有`controller接口`都进行`jwt校验`。 `global`参数通常跟 `unless`一起配合使用。`unless`参数表示需要排除jwt校验的controller接口。

* `global = false`
当`global`值为`false`时, 则表示不进行全局使用，此时我们就需要手动进行引入使用。

* 手动使用

```ts filename="src/controller/JWTController.ts" {5,10}
import { Post, UseBefore } from "@yunflyjs/yunfly";
import { JWTMiddleware, JWTPassThroughMiddleware } from '@yunflyjs/yunfly/build/middleware/JWTMiddleware'

@JsonController('/user')
@UseBefore(JWTMiddleware)  //controller手动使用jwt
export default class JWTController {
  @Inject()  private userService: UserService;
  // 
  @Post('/get-user-info')
  @UseBefore(JWTMiddleware) //具体方法手动使用jwt
  async getUserMsg()Promise<any> {
    return {}
  }
}
```

## passThrough 参数说明

* `passThrough = false`

当 JWT 校验不通过时，直接抛出错误，并阻止后面的程序运行。

* 等同于手动方式的 `JWTMiddleware`

```ts filename="src/controller/JWTController.ts" {4,8}
import { Post, UseBefore } from "@yunflyjs/yunfly";
import { JWTMiddleware } from '@yunflyjs/yunfly/build/middleware/JWTMiddleware'
@JsonController('/user')
@UseBefore(JWTMiddleware)  //controller手动使用jwt
export default class JWTController {
  // 
  @Post('/get-user-info')
  @UseBefore(JWTMiddleware) //具体方法手动使用jwt
  async getUserMsg()Promise<any> {
    return {}
  }
}
```

* `passThrough = true`

当 JWT 校验不通过时，继续执行后面的逻辑, 不阻塞流程。

* 等同于手动方式的 `JWTPassThroughMiddleware`

```ts filename="src/controller/JWTController.ts" {4,8}
import { Post, UseBefore } from "@yunflyjs/yunfly";
import { JWTPassThroughMiddleware } from '@yunflyjs/yunfly/build/middleware/JWTMiddleware'
@JsonController('/user')
@UseBefore(JWTPassThroughMiddleware)  //controller手动使用jwt
export default class JWTController {
  // 
  @Post('/get-user-info')
  @UseBefore(JWTPassThroughMiddleware) //具体方法手动使用jwt
  async getUserMsg()Promise<any> {
    return {}
  }
}
```

## JWT暴露接口说明

* JWT暴露出如下的一些方法 `generateToken`,`verifyToken`,`decodeToken`,`injectToken`,`JWTMiddleware`,`JWTPassThroughMiddleware`

| 方法名 | 说明 |
| ------ | ------ |
| `generateToken`|  根据自定义参数生成jwt秘钥, 例如用户登录时生成秘钥操作 |
| `verifyToken` |  校验jwt秘钥是否有效，自带过期重签功能 |
| `decodeToken` |  解码jwt秘钥，不校验秘钥是否过期，此方法一般不用（慎用） |
| `injectToken` |  生成的jwt秘钥如何处理，可以设置到header头或者cookie中，具体以`config.jwt.token` 配置项为准 |
| `JWTMiddleware` | (config.jwt.global=false) 时使用，手动使用JWT中间件方式，当JWT校验失败是直接抛出错误信息，并中断执行 |
| `JWTPassThroughMiddleware` | (config.jwt.global=true) 时使用，手动使用JWT中间件方式，当JWT校验失败时继续执行 |

* 使用

```ts filename="src/controller/JWTController.ts" {10-17,25}
import { JsonController, Get } from "@yunflyjs/yunfly";
import { generateToken, verifyToken } from '@yunflyjs/yunfly/build/middleware/JWTMiddleware'
//  
@JsonController('/jwt')
export default class JWTController {
  @Get('/gen-token')
  async genToken(
    @Ctx() ctx:Context,
  ): Promise<AnyOptions> {
    const token = generateToken({
      ctx,
      data : {
        name:'zane',
        age: 25,
        hobby:['游泳','爬山']
      },
    })
    return { token }
  }
  //
  @Get('/verify-token')
  async verifyToken(
    @Ctx() ctx: Context,
  ): Promise<AnyOptions> {
    return await verifyToken({ ctx })
  }
}
// verify-token 返回值
// {
//   name:'zane',
//   age: 25,
//   hobby:['游泳','爬山']
// }
```

## 通过 CurrentUser 获得 JWT 数据

```ts filename="src/controller/JWTController.ts" {8,10}
import { JsonController, Get, CurrentUser, UseBefore } from "@yunflyjs/yunfly";
//  
@JsonController('/jwt')
export default class JWTController {
  // 
  @Get('/get-jwt-data')
  async getJWTdata(
    @CurrentUser() user: AnyOptions,
  ): Promise<AnyOptions> {
    const { name, age, hobby } = user || {}
    return { name, age, hobby }
  }
}
// 解码返回值
// {
//   name:'zane',
//   age: 25,
//   hobby:['游泳','爬山']
// }
```

## JWT重签机制 expiredPassThrough

* 值为 `true`, 当 JWT 过期时，会自动重签，也就是说 JWT 不会过期，JWT 仅用来做参数校验。
* 值为 `false`, 当 JWT 过期时，会向外抛出 401 错误，此时 JWT 可用来做 登录态校验 + 参数校验。

### 自动重签规则

> 为了保证服务的正常，当jwt过期时，会拿着过期的密钥进行数据解码，不会阻断当前的request请求（对客户端无影响），但JWT内部会进行过期日志打印和token重签。

* 使用新的token

> * JWT 当前支持 `header头` 和 `cookie` 两种方式
> * 当我们的处理方式为`cookie`类型时，过期会自动重签并设置新的cookie信息（无需做额外处理）。
> * 当我们的处理方式为`header`类型时，过期重签需要进行手动处理。

当 JWT 需要签名或者重签时，请求的`Header头`会返回一个`特殊的重签头标识`: `Set-${key}`, 默认标识为`Set-Authorization`（此处标识取决于你`config.jwt.token.key`字段）,因此我们需要`在客户端统一拦截接口的Header返回头标识`，当发现有`Set-${key}`返回头标识时，取值并`覆盖客户端过期的token即可`。

* 下面以axios为例

### 全局拦截

```js
// ---------------------全局拦截---------------------
// 新增一个全局拦截器
axios.interceptors.response.use(function (response) {
  // 如果返回token为真时，重新替换本地token
  const authorization = response.headers['set-authorization'];
  if(authorization) {
    sessionStorage.setItem('token',authorization);
  }
  return response;
}, function (error) {
  return Promise.reject(error);
});

```

### 单接口拦截

```js
// ----------------------单个接口处理---------------------
axios({
  method: 'post',
  url: 'http://xxx.com/api/xxx',
  data: {},
  headers: {'Authorization': 'xxxxxx'},
}).then((response)=>{
  // 如果返回token为真时，重新替换本地token
  const authorization = response.headers['set-authorization'];
  if(authorization) {
    sessionStorage.setItem('token',authorization);
  }
})
```

* 附加： 如果BFF使用`cors`则需要配置允许的header头字段 `exposeHeaders`(允许返回的header头信息);

```js
config.security = { 
    cors: {
        // 运行返回头
        exposeHeaders: ['Set-Authorization'],
    },
}
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## 1.0.0 (2023-04-24)

### Features

* fast-jwt first version ([39d34f0](https://github.com/yunke-yunfly/yunfly-plugin-fast-jwt/commit/39d34f07fe88e7a6b6d9c86bfa12a458c2df9c29))


src/types.ts
import { Context } from '@yunflyjs/yunfly';

export interface KeysConfig {
  lower: string;
  upper: string;
}

export interface JWTpayload {
  [propsName: string]: any;
}

export interface GenerateTokenOption {
  data: JWTpayload;
  ctx: Context;
}

export interface VerifyTokenOption {
  ctx: Context;
}

export interface DecodeOption {
  json?: boolean;
  complete?: boolean;
}

export interface DecodeTokenOption {
  token: string;
  option?: DecodeOption;
}

export interface InjectTokenOption {
  ctx: Context;
  token: string;
}

export interface JWTOptions {
  enable?: boolean;
  secret?: string;
  expire?: string | number;
  token?: JWTTokenOptions;
  unless?: string[];
  global?: boolean;
  passThrough?: boolean;
  expiredPassThrough?: boolean;
  rsSign?: RsSignOption;
}

export type JWTConfig = JWTOptions | ((ctx: Context) => JWTOptions);

export interface RsSignOption {
  enable?: boolean;
  interval?: number;
}

export interface JWTTokenOptions {
  type?: 'cookie' | 'header';
  key?: string;
  httpOnly?: boolean;
  domain?: string;
  path?: string;
  maxAge?: number;
  expires?: any;
  secure?: boolean;
  sameSite?: boolean | string;
  signed?: boolean;
  overwrite?: boolean;
}


src/utils.ts
import { Context } from 'koa';
import { JWTpayload } from './types';
const cookie = require('cookie');

/**
 *
 * @param {string[]} arr
 * @param {string} str
 * @returns
 */
export const include = (arr: string[], str: string) => {
  if (!arr || !arr.length || !str) {
    return false;
  }

  let result = false;
  for (const key of arr) {
    if (str.indexOf(key) > -1) {
      result = true;
      return result;
    }
  }
  return result;
};

const getValue = (obj: JWTpayload = {}, key: string): string => {
  let res: string = '';
  const keys = Object.keys(obj);
  for (let i = 0; i < keys.length; i++) {
    if (keys[i].toUpperCase() === key) {
      res = obj[keys[i]];
      break;
    }
  }
  return res;
};

/**
 * get jwt token
 *
 * @param {Context} ctx
 * @param {string} type
 * @param {string} key
 */
export const getJwtToken = (ctx: Context, type: string, key: string): string => {
  const headers = ctx.headers || ctx.request.headers || {};
  if (type === 'header') {
    return getValue(headers, key);
  } else if (type === 'cookie') {
    const cookies = cookie.parse(headers.cookie || '');
    return getValue(cookies, key);
  }
  return '';
};


src/middleware/YunflyPluginJWTMiddleware.ts
import {
  UnauthorizedError, BadRequestError, Context,
  Action, KoaMiddlewareInterface, Middleware
} from '@yunflyjs/yunfly';
import ms from 'ms';

const { createSigner, createVerifier, createDecoder } = require('fast-jwt')
const decode = createDecoder();

import {
  DecodeTokenOption,
  GenerateTokenOption,
  InjectTokenOption,
  JWTConfig,
  JWTOptions,
  JWTpayload,
  RsSignOption,
  VerifyTokenOption,
} from '../types';
import { getJwtToken, include } from '../utils';

const DEFAULT_RELOAD_TIME = 900; // 15分钟

export function getJwtConfig(ctx: Context): JWTOptions {
  let jwtConfig: JWTConfig = ctx.config?.jwt || {};
  if (typeof jwtConfig === 'function') {
    jwtConfig = jwtConfig(ctx);
  }
  return jwtConfig;
}

/**
 * 生成token
 *
 * @export
 * @param {GenerateTokenOption} opt
 * @returns
 */
export function generateToken(opt: GenerateTokenOption): string {
  if (!opt.data) throw new UnauthorizedError('加密数据不能为空!');
  const { data, ctx } = opt;
  const JWTConfig = getJwtConfig(ctx);
  if (!JWTConfig.secret) throw new UnauthorizedError('JWT配置secret参数不能为空!');

  if (!JWTConfig.enable) return '';

  const expiresIn = typeof JWTConfig.expire === 'number' ? JWTConfig.expire : ms(JWTConfig.expire as string);
  const signSync = createSigner({ key: JWTConfig.secret, expiresIn })
  const token = signSync(data)

  const resToken = `Bearer ${token}`;

  injectToken({ ctx, token: resToken });

  return resToken;
}

// 判断是否要过期了
export function checkExpired(iat: number, exp: number, rsSign: RsSignOption): boolean {
  if (!rsSign || !rsSign.enable) {
    return false;
  }
  const reloadTime =
    typeof rsSign.interval === 'number' ? rsSign.interval * 60 : DEFAULT_RELOAD_TIME;
  const timestampNow = Math.round(Date.now() / 1000);

  // 有效期小于 重签时间，不处理
  if (exp - iat <= reloadTime) {
    return false;
  }
  // 过期时间小于 重启时间，重签
  if (exp - timestampNow < reloadTime) {
    return true;
  }
  return false;
}

/**
 *
 * 重签 jwt
 * @param {string} token
 * @param {Context} ctx
 */
export function reloadJwt(token: string, ctx: Context): any {
  const data: any = decodeToken({ token });
  delete data.iat;
  delete data.exp;
  ctx.state.payload = data;
  generateToken({ data, ctx });
  return data;
}

/**
 * 验证token
 *
 * @export
 * @param {VerifyTokenOption} opt
 */
export async function verifyToken(opt: VerifyTokenOption): Promise<any> {
  const { ctx } = opt;
  const JWTConfig = getJwtConfigForHttp(ctx);
  if (!JWTConfig.enable) return Promise.resolve({});
  if (!JWTConfig.secret) throw new UnauthorizedError('JWT secret不能为空!');
  if (!JWTConfig.token) throw new UnauthorizedError('JWT token不能为空!');
  const expiredPassThrough = JWTConfig.expiredPassThrough ?? true;
  const token = JWTConfig.token.replace(/Bearer\s?/, '').trim();

  const verifyWithPromise = createVerifier({ key: async () => JWTConfig.secret as string })
  try {
    const decoded = await verifyWithPromise(token);
    // 重签
    if (checkExpired(decoded.iat, decoded.exp, JWTConfig.rsSign as RsSignOption)) {
      return reloadJwt(token, ctx);
    }
    delete decoded.iat;
    delete decoded.exp;
    ctx.state.payload = decoded;
    return decoded;
  } catch (err: any) {
    if (err.message.includes('token has expired')) {
      if (!expiredPassThrough) {
        throw new UnauthorizedError('jwt 验证过期');
      }
      console.info({
        msg: 'jwt 验证过期',
        error: err,
      });
      // jwt 重签
      return reloadJwt(token, ctx);
    }
    throw new UnauthorizedError(err.message || 'jwt token验证失败!');
  }
}

/**
 * 解码token
 *
 * @export
 * @param {DecodeTokenOption} opt
 */
export function decodeToken(opt: DecodeTokenOption) {
  if (!opt.token) throw new UnauthorizedError('jwt token不能为空!');
  return decode(opt.token, opt.option);
}

/**
 * 注入token
 *
 * @export
 * @param {InjectTokenOption} opt
 */
export function injectToken(opt: InjectTokenOption): string {
  const { ctx, token } = opt;
  if (!token) throw new UnauthorizedError('jwt token不能为空!');

  const jwtConfig = getJwtConfig(ctx);
  const tokenConfig = jwtConfig.token || {};
  const key: string = tokenConfig.key || 'Authorization';
  const type: string = tokenConfig.type || 'header';
  const newTokenConfig = { ...tokenConfig };
  if (type === 'header') {
    ctx.set(`Set-${key}`, token);
  } else if (type === 'cookie') {
    const { expires } = tokenConfig || {};
    if (typeof expires === 'function') {
      const exp = expires();
      if (exp) newTokenConfig.expires = exp;
    }
    ctx.cookies.set(key, token, { ...newTokenConfig } as any);
  }
  return token;
}

/**
 * 获得 http 请求的token
 *
 * @param {Context} ctx
 * @returns
 */
function getJwtConfigForHttp(ctx: Context) {
  const jwtConfig = getJwtConfig(ctx);
  const tokenConfig = jwtConfig.token || {};
  const key: string = (tokenConfig.key || 'Authorization').toUpperCase();
  const type: string = tokenConfig.type || 'header';

  const token: string = getJwtToken(ctx, type, key);
  return {
    ...jwtConfig,
    token,
  };
}

/**
 * 用户状态数据
 *
 * @param {Action} action
 * @returns {Promise<AnyOptions>}
 */
export const currentUserChecker = async (action: Action): Promise<JWTpayload> =>
  action.context.state.payload;

/**
 * JWT middleware
 *
 * @param {*} ctx
 * @param {(err?: any) => Promise<any>} next
 * @returns {Promise<any>}
 */
export const JWTMiddleware = async (
  ctx: any,
  next: (err?: any) => Promise<any>,
  passThrough?: boolean,
): Promise<any> => {
  const jwtConfig = getJwtConfig(ctx);

  // 以下条件不做校验
  if (!jwtConfig.enable || !jwtConfig.secret) {
    return await next();
  }

  // jwt 校验
  try {
    await verifyToken({ ctx });
  } catch (err: any) {
    if (!passThrough) {
      throw new BadRequestError(err);
    }
  }

  await next();
};

/**
 * JWT middleware 校验不通过时可继续执行
 *
 * @param {*} ctx
 * @param {(err?: any) => Promise<any>} next
 * @returns {Promise<any>}
 */
export const JWTPassThroughMiddleware = async (
  ctx: any,
  next: (err?: any) => Promise<any>,
): Promise<any> => JWTMiddleware(ctx, next, true);

/**
 * JWT 校验
 *
 * @export
 * @class JWTMiddleware
 * @implements {KoaMiddlewareInterface}
 */
@Middleware({ type: 'before', priority: 18 })
export default class YunflyPluginJWTMiddleware implements KoaMiddlewareInterface {
  async use(ctx: Context, next: (err?: any) => Promise<any>): Promise<any> {
    const jwtConfig = getJwtConfig(ctx);
    // 以下条件不做校验
    if (!jwtConfig.global || (jwtConfig.unless && include(jwtConfig.unless, ctx.request.url))) {
      return await next();
    }
    const passThrough = jwtConfig.passThrough ? true : false;
    return JWTMiddleware(ctx, next, passThrough);
  }
}
